File: r/llikert/R/task.R
#' Define a scoring task
#'
#' A task describes what the model should do with each item: the instructions,
#' the categories with their response codes, optional worked examples, and the
#' prompt format that combines these into the messages the model reads.
#'
#' An **item** is whatever the model responds to: a text to classify (an open
#' survey answer, a post), a questionnaire statement the model answers itself,
#' and so on.
#'
#' @section How the prompt is built:
#' With the default [prompt_format()], the model receives for every item:
#'
#' * a **system** message: your `instructions`, a blank line, `Response
#'   codes:` followed by one line per category (`A = description`, ...), a
#'   blank line, and `Answer with exactly one of the response codes listed
#'   above and nothing else.`
#' * for each example: a **user** message with the example item and an
#'   **assistant** message with its response code;
#' * a **user** message: `Text:`, then the item between `<text>` and `</text>`.
#'
#' The model's probabilities for your response codes are read at the start of
#' its reply. Use [task_messages()] to see the exact messages for any item,
#' and `prompt = prompt_format(...)` to change the wording, the order
#' (instructions before or after the item), how the scale is described, or the
#' answer instruction.
#'
#' Response codes must each be a single token for the service's model;
#' [prepare_task()] checks this before any items are scored.
#'
#' @param name Human-readable task name.
#' @param instructions What you would tell a human coder or respondent. May be
#'   `""` if your prompt format does not use `{instructions}`.
#' @param categories Character vector of category labels, in order.
#' @param responses Character vector of response codes (for example `"A"`,
#'   `"B"`), one per category. Whitespace is meaningful.
#' @param values Optional finite numeric values, one per category, used for the
#'   expected value. An ordinal declaration alone does not supply values.
#' @param ordered Whether the categories are ordinal.
#' @param ids Category ids used for columns and joins. Defaults to the labels,
#'   which must then be unique.
#' @param examples Optional worked examples: a data frame with columns `item`
#'   and `category` (a category id), used in row order.
#' @param prompt A [prompt_format()] describing how the prompt is assembled.
#' @return An object of class `llikert_task`.
#' @export
#' @examples
#' # coding texts
#' task <- scoring_task(
#'   name = "Communicative function",
#'   instructions = "Classify the text's primary communicative function.",
#'   categories = c("description", "question", "request"),
#'   responses = c("A", "B", "C")
#' )
#' task
#' task_messages(task, "Where is the station?")
#'
#' # questionnaire items answered by the model
#' questionnaire <- scoring_task(
#'   name = "Extraversion items",
#'   instructions = "You are completing a personality questionnaire. Rate how well each statement describes you.",
#'   categories = c("disagree strongly", "disagree", "neutral", "agree", "agree strongly"),
#'   responses = c("1", "2", "3", "4", "5"),
#'   values = 1:5,
#'   ordered = TRUE,
#'   prompt = prompt_format(user = "{item}", answer_instruction = "Reply with the number only.")
#' )
#' task_messages(questionnaire, "I am the life of the party.")
scoring_task <- function(name, instructions, categories, responses, values = NULL,
                         ordered = FALSE, ids = categories, examples = NULL, prompt = prompt_format()) {
  check_string(name, "name")
  if (!rlang::is_string(instructions) || is.na(instructions)) {
    llikert_abort("`instructions` must be a single string (it may be empty).", "invalid_task")
  }
  if (!inherits(prompt, "llikert_prompt_format")) llikert_abort("`prompt` must be created with prompt_format().", "invalid_task")
  check_character(categories, "categories")
  check_character(responses, "responses")
  check_character(ids, "ids")
  if (!rlang::is_bool(ordered)) llikert_abort("`ordered` must be TRUE or FALSE.", "invalid_task")
  n <- length(categories)
  if (n < 2L) llikert_abort("A task needs at least two categories.", "invalid_task")
  if (length(responses) != n || length(ids) != n) {
    llikert_abort("`categories`, `responses`, and `ids` must have the same length.", "invalid_task")
  }
  if (!is.null(values)) {
    if (!is.numeric(values) || length(values) != n) {
      llikert_abort("`values` must be numeric with one value per category, or NULL.", "invalid_task")
    }
    if (any(!is.finite(values))) llikert_abort("`values` must all be finite.", "invalid_task")
  }
  cats <- lapply(seq_len(n), function(i) {
    list(id = ids[[i]], label = categories[[i]], response = responses[[i]],
         value = if (is.null(values)) NULL else as.double(values[[i]]))
  })
  exs <- list()
  if (!is.null(examples)) {
    if (!is.data.frame(examples) || !all(c("item", "category") %in% names(examples))) {
      llikert_abort("`examples` must be a data frame with columns `item` and `category`.", "invalid_task")
    }
    exs <- lapply(seq_len(nrow(examples)), function(i) {
      list(item = as.character(examples$item[[i]]), category_id = as.character(examples$category[[i]]))
    })
  }
  new_task(list(schema_version = 1L, name = name, instructions = instructions,
                categories = cats, ordered = ordered, examples = exs, prompt = unclass(prompt)))
}

new_task <- function(x) {
  validate_task(x)
  structure(x, class = "llikert_task")
}

validate_task <- function(x) {
  fail <- function(msg) llikert_abort(msg, "invalid_task", call = rlang::caller_env(2))
  allowed <- c("schema_version", "name", "instructions", "categories", "ordered", "examples", "prompt")
  if (length(setdiff(names(x), allowed))) fail(sprintf("Unknown task fields: %s.", paste(setdiff(names(x), allowed), collapse = ", ")))
  if (!identical(as.integer(x$schema_version), 1L)) fail("Unsupported task schema_version.")
  if (!rlang::is_string(x$name) || !nzchar(x$name) || !valid_utf8(x$name)) fail("`name` must be a nonempty string.")
  if (!rlang::is_string(x$instructions) || !valid_utf8(x$instructions)) fail("`instructions` must be a string.")
  cats <- x$categories
  if (!is.list(cats) || length(cats) < 2L) fail("A task needs at least two categories.")
  for (c in cats) {
    extra <- setdiff(names(c), c("id", "label", "response", "value"))
    if (length(extra)) fail(sprintf("Unknown category fields: %s.", paste(extra, collapse = ", ")))
    for (field in c("id", "label", "response")) {
      if (!rlang::is_string(c[[field]]) || !nzchar(c[[field]]) || !valid_utf8(c[[field]])) {
        fail(sprintf("Category `%s` values must be nonempty strings.", field))
      }
    }
    if (!is.null(c$value) && !(is.numeric(c$value) && length(c$value) == 1L && is.finite(c$value))) {
      fail("Category values must be finite numbers.")
    }
  }
  ids <- vapply(cats, function(c) c$id, character(1))
  if (anyDuplicated(ids)) fail("Category ids must be unique (labels are used as ids unless `ids` is given).")
  if (any(ids %in% c("id", "expected_value"))) fail("Category ids `id` and `expected_value` are reserved.")
  if (anyDuplicated(vapply(cats, function(c) c$response, character(1)))) fail("Response codes must be unique.")
  has_value <- vapply(cats, function(c) !is.null(c$value), logical(1))
  if (any(has_value) && !all(has_value)) fail("Values must be supplied for every category or for none.")
  if (!rlang::is_bool(x$ordered)) fail("`ordered` must be TRUE or FALSE.")
  for (e in x$examples) {
    if (length(setdiff(names(e), c("item", "category_id")))) fail("Examples have the fields `item` and `category_id`.")
    if (!rlang::is_string(e$item) || !nzchar(e$item) || !rlang::is_string(e$category_id)) fail("Examples need a nonempty item and a category id.")
    if (!e$category_id %in% ids) fail(sprintf("Example category `%s` is not a category id.", e$category_id))
  }
  prompt <- x$prompt
  prompt_names <- c("system", "user", "scale", "code", "code_separator", "answer_instruction")
  if (!is.list(prompt) || length(setdiff(names(prompt), prompt_names)) || !all(prompt_names %in% names(prompt))) {
    fail("`prompt` must have the fields system, user, scale, code, code_separator and answer_instruction.")
  }
  for (field in prompt_names) {
    value <- prompt[[field]]
    ok <- if (field == "system") is.null(value) || rlang::is_string(value) else rlang::is_string(value)
    if (!ok || (!is.null(value) && !valid_utf8(value))) fail(sprintf("`prompt$%s` must be a string%s.", field, if (field == "system") " or NULL" else ""))
  }
  validate_prompt_format(prompt, has_values = any(has_value), call = rlang::caller_env(2))
  invisible(x)
}

valid_utf8 <- function(x) all(validUTF8(enc2utf8(x)))

check_string <- function(x, what) {
  if (!rlang::is_string(x) || is.na(x) || !nzchar(x)) llikert_abort(sprintf("`%s` must be a single nonempty string.", what), "invalid_task", call = rlang::caller_env(2))
}

check_character <- function(x, what) {
  if (!is.character(x) || anyNA(x) || any(!nzchar(x))) llikert_abort(sprintf("`%s` must be a character vector without missing or empty values.", what), "invalid_task", call = rlang::caller_env(2))
}

task_as_list <- function(task) {
  out <- unclass(task)
  out$schema_version <- 1L
  out$categories <- lapply(out$categories, function(c) list(id = c$id, label = c$label, response = c$response, value = c$value))
  out$examples <- lapply(out$examples, function(e) list(item = e$item, category_id = e$category_id))
  p <- out$prompt
  out$prompt <- list(system = p$system, user = p$user, scale = p$scale, code = p$code,
                     code_separator = p$code_separator, answer_instruction = p$answer_instruction)
  out[c("schema_version", "name", "instructions", "categories", "ordered", "examples", "prompt")]
}

task_from_list <- function(x) {
  x$ordered <- x$ordered %||% FALSE
  x$examples <- x$examples %||% list()
  defaults <- unclass(prompt_format())
  given <- x$prompt %||% list()
  unknown <- setdiff(names(given), names(defaults))
  if (length(unknown)) llikert_abort(sprintf("Unknown prompt fields: %s.", paste(unknown, collapse = ", ")), "invalid_task")
  # an explicit `"system": null` means no system message; a missing field means the default
  x$prompt <- lapply(stats::setNames(names(defaults), names(defaults)), function(n) if (n %in% names(given)) given[[n]] else defaults[[n]])
  x$categories <- lapply(x$categories, function(c) {
    # `c$value <- NULL` would delete the element; keep an explicit null value
    # integers from JSON become doubles; anything non-numeric is left for validation to reject
    c["value"] <- list(if (is.numeric(c$value)) as.double(c$value) else c$value)
    c
  })
  new_task(x)
}

#' Save and load tasks as portable JSON
#'
#' @param task An `llikert_task`.
#' @param path File path.
#' @return `write_task()` returns `path` invisibly; `read_task()` returns an `llikert_task`.
#' @export
write_task <- function(task, path) {
  stopifnot(inherits(task, "llikert_task"))
  write_json_atomic(task_as_list(task), path)
}

#' @rdname write_task
#' @export
read_task <- function(path) {
  task_from_list(read_json_file(path))
}

#' @export
print.llikert_task <- function(x, ...) {
  cats <- x$categories
  cat_line(cli::format_inline("{.cls llikert_task} {.val {x$name}} ({length(cats)} {if (x$ordered) 'ordered' else 'unordered'} categories)"))
  df <- data.frame(
    id = vapply(cats, function(c) c$id, character(1)),
    response = vapply(cats, function(c) encodeString(c$response, quote = "\""), character(1)),
    value = vapply(cats, function(c) if (is.null(c$value)) "" else format(c$value), character(1)),
    label = vapply(cats, function(c) c$label, character(1)),
    check.names = FALSE
  )
  print(df, row.names = FALSE, right = FALSE)
  if (length(x$examples)) cat_line(cli::format_inline("{length(x$examples)} example{?s}"))
  default <- identical(x$prompt, unclass(prompt_format()))
  cat_line("Prompt format: ", if (default) "default" else "custom", " (see task_messages() for the messages the model receives)")
  invisible(x)
}

cat_line <- function(...) cat(..., "\n", sep = "")

File: python/src/llikert/task.py
"""Portable scoring tasks (task schema v1). The service validates authoritatively; the
checks here fail early, before anything is sent."""

from __future__ import annotations

import json
import math
import os
from dataclasses import asdict, dataclass, field
from typing import Any, Iterable, Mapping, Sequence

from llikert import _prompt

TASK_SCHEMA_VERSION = 1
DEFAULT_ANSWER_INSTRUCTION = "Answer with exactly one of the response codes listed above and nothing else."
RESERVED_CATEGORY_IDS = frozenset({"id", "expected_value"})


def _check_string(value: Any, what: str, allow_empty: bool = False) -> str:
    if not isinstance(value, str):
        raise TypeError(f"{what} must be a string")
    if not allow_empty and value == "":
        raise ValueError(f"{what} must not be empty")
    try:
        value.encode("utf-8")
    except UnicodeEncodeError:
        raise ValueError(f"{what} is not valid Unicode text") from None
    return value


@dataclass(frozen=True)
class Category:
    id: str
    label: str
    response: str
    value: float | None = None

    def __post_init__(self):
        _check_string(self.id, "category id")
        _check_string(self.label, "category label")
        _check_string(self.response, "category response")
        if self.id in RESERVED_CATEGORY_IDS:
            raise ValueError(f"category id {self.id!r} is reserved")
        if self.value is not None:
            if isinstance(self.value, bool) or not isinstance(self.value, (int, float)):
                raise TypeError("category value must be a number")
            if not math.isfinite(self.value):
                raise ValueError("category value must be finite")
            object.__setattr__(self, "value", float(self.value))


@dataclass(frozen=True)
class Example:
    """A worked example shown to the model before each item: an item and its category.

    ``category_id`` is the category's id; :class:`ScoringTask` also accepts a response code
    there and stores the matching id.
    """

    item: str
    category_id: str

    def __post_init__(self):
        _check_string(self.item, "example item")
        _check_string(self.category_id, "example category_id")


@dataclass(frozen=True)
class PromptFormat:
    """How a task and one item become chat messages. See docs/prompts.md.

    Each item produces: a system message from ``system`` (omitted when ``None``), one
    user/assistant pair per example, and a user message from ``user``. Placeholders:

    - ``system`` and ``user``: ``{instructions}``, ``{scale}``, ``{answer_instruction}``
      (the task's ``instructions`` and ``answer_instruction``); ``{item}`` only in
      ``user``, exactly once
    - ``scale``: ``{codes}``, the category lines joined with ``code_separator``
    - ``code``, one line per category: ``{response}`` (required), ``{label}``, ``{value}``, ``{id}``

    Write ``{{`` and ``}}`` for literal braces. Placeholders are filled in one pass, so
    braces inside instructions, labels, or items are never treated as placeholders.
    """

    system: str | None = "{instructions}\n\n{scale}"
    user: str = "Text:\n<text>\n{item}\n</text>\n\n{answer_instruction}"
    scale: str = "Response codes:\n{codes}"
    code: str = "{response} = {label}"
    code_separator: str = "\n"

    def __post_init__(self):
        if self.system is not None:
            _check_string(self.system, "prompt system", allow_empty=True)
        for name in ("user", "scale", "code", "code_separator"):
            _check_string(getattr(self, name), f"prompt {name}", allow_empty=name != "code")
        _prompt.validate({**asdict(self)}, has_values=True)

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)


@dataclass(frozen=True, init=False)
class ScoringTask:
    """A scoring task.

    Build from parallel sequences::

        ScoringTask(name=..., instructions=..., categories=["description", "question"],
                    responses=["A", "B"], values=None, ordered=False)

    ``categories`` are human-readable labels; category ids default to the labels (which
    must then be unique) or are given with ``ids``. Alternatively pass a sequence of
    :class:`Category` records and no ``responses``.
    """

    name: str
    instructions: str
    answer_instruction: str
    categories: tuple[Category, ...]
    ordered: bool
    examples: tuple[Example, ...] = field(default=())
    prompt: PromptFormat = field(default_factory=PromptFormat)
    schema_version: int = TASK_SCHEMA_VERSION

    def __init__(
        self,
        name: str,
        instructions: str,
        categories: Sequence[str] | Sequence[Category],
        responses: Sequence[str] | None = None,
        values: Sequence[float] | None = None,
        ordered: bool = False,
        ids: Sequence[str] | None = None,
        examples: Iterable[Example | tuple[str, str] | Mapping[str, str]] | None = None,
        answer_instruction: str = DEFAULT_ANSWER_INSTRUCTION,
        prompt: PromptFormat | None = None,
    ):
        _check_string(name, "name")
        _check_string(instructions, "instructions", allow_empty=True)
        _check_string(answer_instruction, "answer_instruction", allow_empty=True)
        if not isinstance(ordered, bool):
            raise TypeError("ordered must be True or False")
        categories = list(categories)
        if categories and all(isinstance(c, Category) for c in categories):
            if responses is not None or values is not None or ids is not None:
                raise ValueError("pass either Category records or labels with responses/values/ids, not both")
            records = categories
        else:
            if responses is None:
                raise ValueError("responses are required when categories are labels")
            labels = categories
            responses = list(responses)
            ids = list(ids) if ids is not None else labels
            if values is not None:
                values = list(values)
            lengths = {len(labels), len(responses), len(ids)} | ({len(values)} if values is not None else set())
            if len(lengths) != 1:
                raise ValueError("categories, responses, ids and values must have the same length")
            records = [
                Category(id=i, label=label, response=r, value=None if values is None else v)
                for i, label, r, v in zip(ids, labels, responses, values if values is not None else [None] * len(labels))
            ]
        if len(records) < 2:
            raise ValueError("a task needs at least two categories")
        if len({c.id for c in records}) != len(records):
            raise ValueError("category ids must be unique (labels are used as ids unless ids are given)")
        if len({c.response for c in records}) != len(records):
            raise ValueError("response strings must be unique")
        if len({c.value is None for c in records}) != 1:
            raise ValueError("values must be supplied for every category or for none")

        parsed_examples = []
        for ex in examples or ():
            if isinstance(ex, Example):
                parsed_examples.append(ex)
            elif isinstance(ex, Mapping):
                parsed_examples.append(Example(item=ex["item"], category_id=ex["category_id"]))
            else:
                item, category_id = ex
                parsed_examples.append(Example(item=item, category_id=category_id))
        parsed_examples = [_resolve_example(ex, records) for ex in parsed_examples]

        prompt = prompt if prompt is not None else PromptFormat()
        if not isinstance(prompt, PromptFormat):
            raise TypeError("prompt must be a PromptFormat")
        if "value" in _prompt.fields(_prompt.parse_template(prompt.code, _prompt.CODE_FIELDS, "prompt.code")) and records[0].value is None:
            raise ValueError("prompt.code uses {value}, but the categories have no values")

        object.__setattr__(self, "name", name)
        object.__setattr__(self, "instructions", instructions)
        object.__setattr__(self, "answer_instruction", answer_instruction)
        object.__setattr__(self, "categories", tuple(records))
        object.__setattr__(self, "ordered", ordered)
        object.__setattr__(self, "examples", tuple(parsed_examples))
        object.__setattr__(self, "prompt", prompt)
        object.__setattr__(self, "schema_version", TASK_SCHEMA_VERSION)

    @property
    def has_values(self) -> bool:
        return self.categories[0].value is not None

    def to_dict(self) -> dict[str, Any]:
        return {
            "schema_version": self.schema_version,
            "name": self.name,
            "instructions": self.instructions,
            "answer_instruction": self.answer_instruction,
            "categories": [{"id": c.id, "label": c.label, "response": c.response, "value": c.value} for c in self.categories],
            "ordered": self.ordered,
            "examples": [{"item": e.item, "category_id": e.category_id} for e in self.examples],
            "prompt": self.prompt.to_dict(),
        }

    def messages(self, item: str) -> list[dict[str, str]]:
        """The chat messages the model receives for ``item``, built locally (no service needed).

        The model then sees these messages in its own chat format; ``PreparedTask.preview_prompt()``
        shows that exact text after preparation.
        """
        _check_string(item, "item", allow_empty=True)
        return _prompt.build_messages(self.to_dict(), item)

    @classmethod
    def from_dict(cls, obj: Mapping[str, Any]) -> ScoringTask:
        allowed = {"schema_version", "name", "instructions", "answer_instruction", "categories", "ordered", "examples", "prompt"}
        unknown = set(obj) - allowed
        if unknown:
            raise ValueError(f"unknown task fields: {sorted(unknown)}")
        if obj.get("schema_version") != TASK_SCHEMA_VERSION:
            raise ValueError(f"unsupported task schema_version {obj.get('schema_version')!r}")
        records = []
        for c in obj.get("categories", []):
            unknown = set(c) - {"id", "label", "response", "value"}
            if unknown:
                raise ValueError(f"unknown category fields: {sorted(unknown)}")
            records.append(Category(id=c["id"], label=c["label"], response=c["response"], value=c.get("value")))
        return cls(
            name=obj.get("name"),
            instructions=obj.get("instructions"),
            categories=records,
            ordered=obj.get("ordered", False),
            examples=obj.get("examples", []),
            answer_instruction=obj.get("answer_instruction", DEFAULT_ANSWER_INSTRUCTION),
            prompt=_prompt_from_dict(obj.get("prompt")),
        )

    def to_json(self, path: str | os.PathLike) -> None:
        with open(path, "w", encoding="utf-8") as f:
            json.dump(self.to_dict(), f, indent=2, ensure_ascii=False, allow_nan=False)
            f.write("\n")

    @classmethod
    def from_json(cls, path: str | os.PathLike) -> ScoringTask:
        with open(path, encoding="utf-8") as f:
            return cls.from_dict(json.load(f))


def _resolve_example(example: Example, records: list[Category]) -> Example:
    """An example names its category by id; a response code is accepted too, because that is
    what the example shows the model. Ids win when a value is both an id and another code."""
    ids = [c.id for c in records]
    if example.category_id in ids:
        return example
    matches = [c.id for c in records if c.response == example.category_id]
    if len(matches) == 1:
        return Example(item=example.item, category_id=matches[0])
    raise ValueError(
        f"example category {example.category_id!r} is not one of this task's categories; "
        f"use a category id ({', '.join(repr(i) for i in ids)}) "
        f"or a response code ({', '.join(repr(c.response) for c in records)})"
    )


def _prompt_from_dict(obj: Mapping[str, Any] | None) -> PromptFormat:
    if obj is None:
        return PromptFormat()
    unknown = set(obj) - {"system", "user", "scale", "code", "code_separator"}
    if unknown:
        raise ValueError(f"unknown prompt fields: {sorted(unknown)}")
    return PromptFormat(**obj)
